File: examples/ag-ui/servers/go/messages.go
package main

import (
	"encoding/json"
	"strings"
)

func textFromMessage(msg incomingMessage) string {
	if len(msg.Parts) > 0 {
		var parts []string
		for _, part := range msg.Parts {
			if part.Type == "text" && part.Content != "" {
				parts = append(parts, part.Content)
			}
		}
		if len(parts) > 0 {
			return strings.Join(parts, "")
		}
	}

	if len(msg.Content) == 0 {
		return ""
	}

	var asString string
	if err := json.Unmarshal(msg.Content, &asString); err == nil {
		return asString
	}

	var asArray []map[string]any
	if err := json.Unmarshal(msg.Content, &asArray); err == nil {
		var parts []string
		for _, item := range asArray {
			if item["type"] == "text" {
				if text, ok := item["text"].(string); ok {
					parts = append(parts, text)
				}
			}
		}
		return strings.Join(parts, "")
	}

	return strings.TrimSpace(string(msg.Content))
}

func toChatMessages(messages []incomingMessage) (system string, chat []chatMessage) {
	for _, msg := range messages {
		role := msg.Role
		switch role {
		case "developer":
			role = "system"
		case "tool", "reasoning":
			continue
		}

		text := textFromMessage(msg)
		if text == "" {
			continue
		}

		switch role {
		case "system":
			if system != "" {
				system += "\n\n"
			}
			system += text
		case "user", "assistant":
			chat = append(chat, chatMessage{Role: role, Content: text})
		}
	}

	return system, chat
}

File: examples/ag-ui/servers/rust/src/messages.rs
use crate::agui::{ChatMessage, IncomingMessage};

pub fn to_chat_messages(messages: &[IncomingMessage]) -> (String, Vec<ChatMessage>) {
    let mut system = String::new();
    let mut chat = Vec::new();

    for message in messages {
        let mut role = message.role.as_str();
        match role {
            "developer" => role = "system",
            "tool" | "reasoning" => continue,
            _ => {}
        }

        let text = text_from_message(message);
        if text.is_empty() {
            continue;
        }

        match role {
            "system" => {
                if !system.is_empty() {
                    system.push_str("\n\n");
                }
                system.push_str(&text);
            }
            "user" | "assistant" => chat.push(ChatMessage {
                role: role.to_string(),
                content: text,
            }),
            _ => {}
        }
    }

    (system, chat)
}

fn text_from_message(message: &IncomingMessage) -> String {
    if let Some(parts) = &message.parts {
        let joined = parts
            .iter()
            .filter(|part| part.part_type == "text")
            .filter_map(|part| part.content.as_deref())
            .collect::<Vec<_>>()
            .join("");
        if !joined.is_empty() {
            return joined;
        }
    }

    let Some(content) = &message.content else {
        return String::new();
    };

    if let Some(text) = content.as_str() {
        return text.to_string();
    }

    if let Some(items) = content.as_array() {
        let joined = items
            .iter()
            .filter_map(|item| {
                if item.get("type")?.as_str()? == "text" {
                    item.get("text")?.as_str().map(str::to_string)
                } else {
                    None
                }
            })
            .collect::<Vec<_>>()
            .join("");
        if !joined.is_empty() {
            return joined;
        }
    }

    content.to_string()
}
